### PR TITLE
Dequeue support

### PR DESCRIPTION
**Pull request by jpowerwa**
Add getResults and getStatusAndResult to scala interface and get_results and get_status_and_result to python interface. These methods are intended to facilitate dequeue logic.

### CODE REVIEW

File: python/papertasks.py [hunk 1]
```diff
@@ -127,7 +132,61 @@ def get_paper_ids(self, model_version: int, paper_id_range_start: str=None, pape
...
+
+        try:
+            with self.conn.cursor() as cur:
+                cur.execute(sql, (model_version, paper_id))
+                record = cur.fetchone()
+        except:
+            self.conn.rollback()
+            raise
+
+        return [record[0], record[1]] if record else None
```
Comment (dirkgr) [inline python/papertasks.py]: It says it returns a `Tuple`, but this is a `List`.

File: scala/src/main/scala/org/allenai/s2/papertasks/TaskDB.scala [hunk 2]
```diff
@@ -106,18 +116,50 @@ class TaskDB(
...
 
+  /* Get results for paper_ids in optionally specified range with status='Done',
+   * optionally updated since specified datetime.
+   * Paper id range is inclusive of start and end range ids.
+   */
+  def getResults(modelVersion: Short, statusChangedSince: Option[LocalDateTime],
+    paperIdRangeStart: Option[String], paperIdRangeEnd: Option[String]): Seq[Tuple2[String, JsValue]] = {
+    val start = paperIdRangeStart.getOrElse(MIN_PAPER_ID)
+    val end = paperIdRangeEnd.getOrElse(MAX_PAPER_ID)
+    val sinceDate = statusChangedSince.getOrElse(LocalDateTime.of(2016, 1, 1, 0, 0))
```
Comment (dirkgr) [inline scala/src/main/scala/org/allenai/s2/papertasks/TaskDB.scala]: Is that really the earliest date we have?

File: scala/src/main/scala/org/allenai/s2/papertasks/TaskDB.scala [hunk 3]
```diff
@@ -106,18 +116,50 @@ class TaskDB(
...
             AND paperid BETWEEN $start AND $end""".map { rs => rs.string(1) }.list().apply()
     }
   }
 
+  /* Get results for paper_ids in optionally specified range with status='Done',
+   * optionally updated since specified datetime.
+   * Paper id range is inclusive of start and end range ids.
+   */
+  def getResults(modelVersion: Short, statusChangedSince: Option[LocalDateTime],
+    paperIdRangeStart: Option[String], paperIdRangeEnd: Option[String]): Seq[Tuple2[String, JsValue]] = {
```
Comment (dirkgr) [inline scala/src/main/scala/org/allenai/s2/papertasks/TaskDB.scala]: You can write that type annotation as `Seq[(String, JsValue)]`.

Comment (dirkgr) [reply scala/src/main/scala/org/allenai/s2/papertasks/TaskDB.scala]: All of those `Options` could default to `None`, but I guess you don't need that?

File: scala/src/main/scala/org/allenai/s2/papertasks/TaskDB.scala [hunk 4]
```diff
@@ -106,18 +116,50 @@ class TaskDB(
...
+    val start = paperIdRangeStart.getOrElse(MIN_PAPER_ID)
+    val end = paperIdRangeEnd.getOrElse(MAX_PAPER_ID)
+    val sinceDate = statusChangedSince.getOrElse(LocalDateTime.of(2016, 1, 1, 0, 0))
+
+    localTx { implicit session =>
+      sql"""SELECT paperid, result FROM tasks
+            WHERE modelversion = $modelVersion
+            AND status = 'Done'::processing_status
+            AND statusChanged > $sinceDate
+            AND paperid BETWEEN $start AND $end""".map { rs => Tuple2(rs.string(1), asJsValue(rs.string(columnIndex=2))) }.list().apply()
```
Comment (dirkgr) [inline scala/src/main/scala/org/allenai/s2/papertasks/TaskDB.scala]: Same thing here. You don't have to write `Tuple2`. Just write `(x, y)`.

File: scala/src/main/scala/org/allenai/s2/papertasks/TaskDB.scala (hunk 2 above)
Comment (jpowerwa) [reply scala/src/main/scala/org/allenai/s2/papertasks/TaskDB.scala]: Well, the papertasks github repo came into being in April of 2018, so I figured going back two years was safe.

File: scala/src/main/scala/org/allenai/s2/papertasks/TaskDB.scala (hunk 4 above)
Comment (jpowerwa) [reply scala/src/main/scala/org/allenai/s2/papertasks/TaskDB.scala]: Ah. Much nicer. Thanks for that.

File: python/papertasks.py (hunk 1 above)
Comment (jpowerwa) [reply python/papertasks.py]: Fixed.

**Comment by jpowerwa** [comment]
Thanks for the review, @dirkgr. I think I've fixed everything you commented on. Thanks for the scala tips, as usual.

File: scala/src/main/scala/org/allenai/s2/papertasks/TaskDB.scala (hunk 2 above)
Comment (dirkgr) [reply scala/src/main/scala/org/allenai/s2/papertasks/TaskDB.scala]: Why not pick 1/1/1970?

Comment (jpowerwa) [reply scala/src/main/scala/org/allenai/s2/papertasks/TaskDB.scala]: Fair enough.

**Pull request merged**